### DISCUSSION TITLE
useFetcher - only some requests are cancelled after interuptions

### DISCUSSION THREAD

**Discussion by jmlee2k** [remix-run/remix#4540] [Q&A]
I've just started exploring Remix, so I fully understand that I may have got this wrong, but I expected that Remix would cancel *all* stale fetch requests when initiating a new one.

I've created a simple test case to show the issue:
https://stackblitz.com/edit/node-yp8gvx?file=app%2Froutes%2Findex.tsx,app%2Froutes%2Fchars.tsx

It uses `useFetcher` to make a call to resource route every time the text in the input box changes.  I've added an artificial 4s delay to the loader on the resource route to clearly show what's happening.

When I type "homer" in the search box, 5 search requests are sent, which makes sense:

1. "h"
2. "ho"
3. "hom"
4. "home"
5. "homer"

I would expect that the first 4 fetches would be cancelled, since the following one would interrupt it (assuming all were typed within the 4s delay), however, it looks only some of the requests are cancelled:

![image](https://user-images.githubusercontent.com/7396545/200431956-cfcf5a63-bf4e-492e-9d99-54342011d9a1.png)

If I type "this is a test", it looks like every other request is cancelled:

![image](https://user-images.githubusercontent.com/7396545/200432225-c48063c0-1f23-42cc-ac94-e75f200fe644.png)

Am I missing something here?

Thanks in advance!

**Comment by machour** [comment]
Is your backend fulfilling those requests ? (We're missing an interesting part of your screenshot with requests times)

**Reply by jmlee2k** [reply]
I've made a [repo ](https://github.com/jmlee2k/remix-usefetcher-test) with the test case and ran it both locally and in stackblitz.

I've included both the chrome devtools screenshots and the dev server output.  The results are slightly different (I assume due to the fact that stackblitz runs via a service worker), but both show that some stale requests aren't cancelled.

## Local
![image](https://user-images.githubusercontent.com/7396545/200595972-66f5709f-d55b-40de-9c8f-46c6561e4650.png)
```
GET /search?q=h&_data=routes%2Fsearch - - - - ms
GET /search?q=hom&_data=routes%2Fsearch - - - - ms
GET /search?q=ho&_data=routes%2Fsearch 200 - - 4007.767 ms
GET /search?q=home&_data=routes%2Fsearch 200 - - 4014.332 ms
GET /search?q=homer&_data=routes%2Fsearch 200 - - 4017.660 ms
```

## Stackblitz
https://stackblitz.com/github/jmlee2k/remix-usefetcher-test

![image](https://user-images.githubusercontent.com/7396545/200596192-31dddc13-d6ee-4be7-9d3a-451ab36af25d.png)
```
GET /search?q=h&_data=routes%2Fsearch 200 - - 4009.955 ms
GET /search?q=ho&_data=routes%2Fsearch 200 - - 4010.095 ms
GET /search?q=hom&_data=routes%2Fsearch 200 - - 4003.055 ms
GET /search?q=home&_data=routes%2Fsearch 200 - - 4010.225 ms
GET /search?q=homer&_data=routes%2Fsearch 200 - - 4010.725 ms
```

**Comment by jmlee2k** [comment]
I've also tested this using @kentcdodds informative [full stack components](https://www.epicweb.dev/full-stack-components) article (which is essentially a fully-realized version of this test case), and I'm getting similar results after adding an artificial delay.

![image](https://user-images.githubusercontent.com/7396545/200626425-47b0a934-0841-4c84-ac83-beee130984b9.png)

**Reply by melodyclue** [reply]
Have you found a solution?
